tools/flaky-test-reporter/slack_notification.go

```diff
@@ -0,0 +1,170 @@
+func newSlackClient(slackTokenPath string) (*SlackClient, error) {
+	b, err := ioutil.ReadFile(slackTokenPath)
+	if err != nil {
+		return nil, err
+	}
+	return &SlackClient{
+		userName: knativeBotName,
+		tokenStr: strings.TrimSpace(string(b)),
+	}, nil
+}
+
+// postMessage does http post
+func (c *SlackClient) postMessage(url string, uv *url.Values) error {
+	resp, err := http.PostForm(url, *uv)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	t, _ := ioutil.ReadAll(resp.Body)
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("http response code is not 200 '%s'", string(t))
+	}
+	// response code could also be 200 if channel doesn't exist, parse response body to find out
+	var b struct {
+		OK bool `json:"ok"`
+	}
+	if err = json.Unmarshal(t, &b); nil != err || !b.OK {
+		return fmt.Errorf("response not ok '%s'", string(t))
+	}
+	return nil
+}
+
+// writeSlackMessage posts text to channel
+func (c *SlackClient) writeSlackMessage(text, channel string) error {
+	uv := &url.Values{}
+	uv.Add("username", c.userName)
+	uv.Add("token", c.tokenStr)
+	if nil != c.iconEmoji {
+		uv.Add("icon_emoji", *c.iconEmoji)
+	}
+	uv.Add("channel", channel)
+	uv.Add("text", text)
+
+	return c.postMessage(slackChatPostMessageURL, uv)
+}
+
+// getTestgridTabURL gets Testgrid URL for giving job
+func getTestgridTabURL(jobName string) (string, error) {
+	url, ok := jobNameTestgridURLMap[jobName]
+	if !ok {
+		return "", fmt.Errorf("cannot find Testgrid tab for job '%s'", jobName)
+	}
+	return fmt.Sprintf("%s/%s&%s", testgrid.BaseURL, url, testgridFilter), nil
+}
+
+// createSlackMessageForRepo creates slack message layout from RepoData
+func createSlackMessageForRepo(rd *RepoData) string {
+	flakyTests := getFlakyTests(rd)
+	message := fmt.Sprintf("(As of %s)\nActive flaky tests in repo '%s': %d",
+		time.Unix(*rd.LastBuildStartTime, 0).String(), rd.Config.Repo, len(flakyTests))
+	for _, testName := range flakyTests {
+		message += fmt.Sprintf("\n>- %s", testName)
+		// TODO(chaodaiG): try adding Github issues when PR #506 merged
+	}
+	if testgirdTabURL, err := getTestgridTabURL(rd.Config.Name); nil != err {
+		log.Println(err) // don't fail as this could be optional
+	} else {
+		message += fmt.Sprintf("\nSee Testgrid tab for most recent flaky tests: %s", testgirdTabURL)
+	}
+	return message
+}
+
+func sendSlackNotifications(repoDataAll []*RepoData, c *SlackClient, dryrun *bool) error {
+	var allErrs []error
+	for _, rd := range repoDataAll {
+		channels, ok := slackChannelsMap[rd.Config.Repo]
+		if !ok {
+			log.Printf("cannot find Slack channel for repo '%s', skip Slack notifiction", rd.Config.Repo)
+			continue
+		}
+
+		ch := make(chan bool, len(channels))
+		wg := sync.WaitGroup{}
+		for _, channel := range channels {
+			wg.Add(1)
+			go func(wg *sync.WaitGroup) {
+				message := createSlackMessageForRepo(rd)
+				if err := run(
+					fmt.Sprintf("post Slack message for repo '%s' in channel '%s'", rd.Config.Repo, channel.name),
```

Comment:
Debug?

Reply:
not really debug, the `run` help function is a wrapper for dryrun, which prints this message with `[dryrun]` prefix if in dryrun mode, otherwise print this message straight